### PR TITLE
recursive-patterns(20) Add language version 8.0 for recursive pattern-matching.

### PR DESCRIPTION
**Pull request by gafter**
@cston @agocke Please review.
@dotnet/roslyn-compiler FYI

### CODE REVIEW

**Comment by agocke** [comment]
LGTM, but @jaredpar I think this is the canary for needing a proper dev16 branch. These kinds of changes will be harder to merge if we don't have a shared 8.0 branch to put them in.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs [hunk 1]
```diff
@@ -28,7 +28,6 @@ public void DiagnosticAnalyzerAllInOne()
             var syntaxKinds = new HashSet<SyntaxKind>();
             syntaxKinds.Add(SyntaxKind.SubpatternElement);
             syntaxKinds.Add(SyntaxKind.DeconstructionPattern);
-            syntaxKinds.Add(SyntaxKind.DiscardPattern);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.AllInOne.cs]: Is this change intentional?

File: src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs [hunk 2]
```diff
@@ -194,6 +194,21 @@ public class Program
...
+            await TestLanguageVersionUpgradedAsync(
+$@"
+class Program
+{{
+#error version:[|{LanguageVersion.Latest.MapSpecifiedToEffectiveVersion().ToDisplayString()}|]
+}}",
+                LanguageVersion.Latest.MapSpecifiedToEffectiveVersion(),
+                new CSharpParseOptions(LanguageVersion.CSharp7_3));
+        }
+        #endregion C# 8.0
```
Comment (cston) [inline src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs]: Move below `#region C# 7.3`.
